### PR TITLE
Allow multiple submissions in modeling exercises

### PR DESCRIPTION
**Pull request by maxr96**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

We want to allow students to submit exercise until the exercise deadline.
The are following criteria for allowing to submit an exercise are the following:
    - There is no due date
    - The due date is before the current date
    - The participation initialization date is after the due date
    - The result is not yet set (avoid infinite submission for late and without due date submissions)

### Description

The Save button is gone, but the student can submit as much as needed and last submission before the exercise due date will be reviewed. If student started participating in the exercise after the due date the submit button will be called `Submit(deadline missed)`. Tooltips added for both buttons. 

### Steps for Testing

1. Log in to Artemis
2. Navigate to Overview
3. Select Course with Modelling Exercise
4. Start new Modeling Exercise
5. Submit exercise multiple time
6. Wait for exercise deadline and check that Submit button is disabled now

### Screenshots

Peek 2019-11-09 15-27

### CODE REVIEW

**Comment by krusche** [comment]
Auto-save should still invoke the save REST call and NOT the submit call.
Please double check the transition to participation.initializationState = Finished.

File: src/main/webapp/app/modeling-submission/modeling-submission.component.ts [hunk 1]
```diff
@@ -224,106 +229,73 @@ export class ModelingSubmissionComponent implements OnInit, OnDestroy, Component
...
                 },
-                err => {
+                () => {
                     this.jhiAlertService.error('artemisApp.modelingEditor.error');
                     this.submission.submitted = false;
                 },
+                () => (this.isSaving = false),
+            );
+        } else {
+            this.modelingSubmissionService.create(this.submission, this.modelingExercise.id).subscribe(
```
Comment (krusche) [inline src/main/webapp/app/modeling-submission/modeling-submission.component.ts]: why did you delete update here?
I think we should always use update instead of create

File: src/main/webapp/i18n/de/global.json [hunk 2]
```diff
@@ -198,7 +198,10 @@
             "registerForCourse": "Für Kurs anmelden",
             "exercise": "Aufgaben",
             "scores": "Bewertungen",
-            "manageHints": "Verwalte Hinweise"
+            "manageHints": "Verwalte Hinweise",
+            "submitTooltip": "Du kannst mehrmals abgeben. Deine Bewertung ergibt sich immer aus der letzten Abgabe.",
```
Comment (krusche) [inline src/main/webapp/i18n/de/global.json]: include "vor der Abgabefrist"
also use "Abgabefrist" instead of "Deadline"

File: src/main/webapp/i18n/en/global.json [hunk 3]
```diff
@@ -204,7 +205,10 @@
             "registerForCourse": "Register For Course",
             "exercise": "Exercises",
             "scores": "Scores",
-            "manageHints": "Manage Hints"
+            "manageHints": "Manage Hints",
+            "submitTooltip": "You can submit multiple times. Your latest submission will count towards your score.",
```
Comment (krusche) [inline src/main/webapp/i18n/en/global.json]: "before the deadline"

File: src/test/java/de/tum/in/www1/artemis/ModelingAssessmentIntegrationTest.java [hunk 4]
```diff
@@ -376,20 +376,6 @@ public void automaticAssessmentUponModelSubmission_noResultInDatabase() throws E
         assertThat(automaticResult).as("automatic result not stored in database").isNotPresent();
     }
 
-    @Test
-    @WithMockUser(username = "student2")
-    public void noAutomaticAssessmentUponModelSave() throws Exception {
-        saveModelingSubmissionAndAssessment(true);
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/ModelingAssessmentIntegrationTest.java]: why do you delete this test case? It is still valid!

File: src/main/webapp/app/modeling-submission/modeling-submission.component.ts [hunk 5]
```diff
@@ -494,4 +466,11 @@ export class ModelingSubmissionComponent implements OnInit, OnDestroy, Component
         this.showComplaintForm = false;
         this.showRequestMoreFeedbackForm = !this.showRequestMoreFeedbackForm;
     }
+
+    /**
+     * The exercise is still active if it's due date hasn't passed yet, or if the participation was started after the due date has already passed
+     */
+    get isActive() {
+        return this.modelingExercise && (!this.modelingExercise.dueDate || moment(this.modelingExercise.dueDate).isAfter(moment()));
```
[the remark this answers is not in the document]
Comment (maxr96) [reply src/main/webapp/app/modeling-submission/modeling-submission.component.ts]: Yes but I do it with isLate property. So if the exercise was started after the due date, student will see `late submission` button

File: src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java [hunk 6]
```diff
@@ -204,15 +204,9 @@ public ModelingSubmission save(ModelingSubmission modelingSubmission, ModelingEx
...
-        // a chance to try it out again.
-        // TODO: think about how we can enable retry again in the future in a fair way
-        // make sure that no (submitted) submission exists for the given user and exercise to prevent retry submissions
-        boolean submittedSubmissionExists = participation.getSubmissions().stream().anyMatch(Submission::isSubmitted);
-        if (submittedSubmissionExists) {
-            throw new BadRequestAlertException("User " + username + " already participated in exercise with id " + modelingExercise.getId(), "modelingSubmission",
-                    "participationExists");
+        final var exerciseDueDate = modelingExercise.getDueDate();
+        if (exerciseDueDate != null && exerciseDueDate.isBefore(ZonedDateTime.now()) && participation.getInitializationDate().isBefore(exerciseDueDate)) {
+            throw new BadRequestAlertException("You missed the deadline for submitting modeling exercise", "modelingSubmission", "idexists");
```
[the remark this answers is not in the document]
Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java]: Sure

File: src/main/webapp/app/modeling-submission/modeling-submission.component.ts (hunk 1 above)
Comment (maxr96) [reply src/main/webapp/app/modeling-submission/modeling-submission.component.ts]: I didn't delete it. I only check if we already have submission, then we update it, if not then create.

**Comment by maxr96** [comment]
> Good job implementing this feature @maxr96
> I have tested it and the functionality and requests that are sent looked good to me.
> 
> However there is still an error that is thrown when you initially open the apollon-editor
> image

Thanks for checking @lovaiible. Seems like change detection where triggered multiple times before we actually get the modeling exercise. I added an explicit check to avoid receiving the error

**Comment by maxr96** [comment]
> Thanks for the help. Looks good to me! You should only have a look at you `isActive` property again

Thanks for checking. I looked again at `isActive`, but for me it covered all the cases. I added a note however about the result that we don't check

**Comment by maxr96** [comment]
> Auto-save should still invoke the save REST call and NOT the submit call.
> Please double check the transition to participation.initializationState = Finished.

@krusche I see, I returned the saving functionality and auto-save calls saving endpoint again. Also since the saving is relevant again I added back the removed integration test. 
Regarding the `participation.initializationState = Finished` transition I didn't notice any strange behavior. After the student click submit he will only see `Submitted, waiting for grading.` message on page reload.

File: src/main/webapp/app/modeling-submission/modeling-submission.component.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ungaralex) [reply src/main/webapp/app/modeling-submission/modeling-submission.component.ts]: From our Slack discussion: 
> except the participation start date is after the exercise due date, then they can submit forever

Also, please change the comment then, because there you say that `isActive` checks if the participation was started after the due date. It's fine if you use the `isLate` property, but then the JSDoc should reflect that.

[the remark this answers is not in the document]
Comment (maxr96) [reply src/main/webapp/app/modeling-submission/modeling-submission.component.ts]: @ungaralex True, I forgot to update JSDoc after the latest changes.

File: src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java [hunk 7]
```diff
@@ -204,15 +205,9 @@ public ModelingSubmission save(ModelingSubmission modelingSubmission, ModelingEx
...
-        // get the manual feedback early and can then retry the exercise within the deadline and have a second chance, others might get the manual feedback late and would not have
-        // a chance to try it out again.
-        // TODO: think about how we can enable retry again in the future in a fair way
-        // make sure that no (submitted) submission exists for the given user and exercise to prevent retry submissions
-        boolean submittedSubmissionExists = participation.getSubmissions().stream().anyMatch(Submission::isSubmitted);
-        if (submittedSubmissionExists) {
-            throw new BadRequestAlertException("User " + username + " already participated in exercise with id " + modelingExercise.getId(), "modelingSubmission",
-                    "participationExists");
+        final var exerciseDueDate = modelingExercise.getDueDate();
+        if (exerciseDueDate != null && exerciseDueDate.isBefore(ZonedDateTime.now()) && participation.getInitializationDate().isBefore(exerciseDueDate)) {
```
[the remark this answers is not in the document]
Comment (maxr96) [reply src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java]: You are totally right here! Didn't think about this

File: src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java [hunk 8]
```diff
@@ -401,6 +409,39 @@ public void getModelSubmissionForModelingEditor() throws Exception {
...
+    @Test
+    @WithMockUser(value = "student3", roles = "USER")
+    public void submitExercise_beforeDueDateSecondSubmission_allowed() throws Exception {
+        submittedSubmission.setModel(emptyModel);
+        submittedSubmission = request.putWithResponseBody("/api/exercises/" + classExercise.getId() + "/modeling-submissions", submittedSubmission, ModelingSubmission.class,
+                HttpStatus.OK);
+
+        final var submissionInDb = modelingSubmissionRepo.findById(submittedSubmission.getId());
+        assertThat(submissionInDb.isPresent());
+        assertThat(submissionInDb.get().getModel()).isEqualTo(emptyModel);
```
[the remark this answers is not in the document]
Comment (maxr96) [reply src/test/java/de/tum/in/www1/artemis/ModelingSubmissionIntegrationTest.java]: True!

**Comment by maxr96** [comment]
> I reviewed your changes and it looks good to me! I have added some minor comments, can you have a look at them?

Thank you for checking @lovaiible . I implemented the suggested changes

**Comment by maxr96** [comment]
>     * It still says, I have unsaved changes (alert) even after submitting and trying to leave the page.
> 
>     * It also says late submission in the upper right corner above the submit button, although the deadline hasn't expired, yet.
> 
> 
> Apart from that, the rest looks fine +1

Thank you for checking my PR.
1. You are right. Had to debug it a bit to find out that the size was saved differently when reassigning the model on updating the submission. Not it should work as expected.
2. I first checked and it was only saying late submission if the assessment due date is passed or it is null. But after submitting I also had this strange behavior and I figured out that we are missing submissions on the participation. I reconnected them and now the string is the same as on course-exercise page.

**Comment by krusche** [comment]
Can you please update the description and add a screenshot / screencast?

**Comment by maxr96** [comment]
@krusche I updated the description. Added a screen cast showing multiple submission in action. Also added multiple client tests to check if the submission is allowed or blocked properly

**Comment by krusche** [comment]
@maxr96 can you please resolve the merge conflicts?
